OCaml_Intro_to_CS/part_8.ml

```diff
@@ -0,0 +1,189 @@
+    | Node (l, x, r) -> Node (aux l, f x, aux r)
+in
+aux tree
+;;
+
+let show_tree tree =
+  let rec aux = function
+    | Leaf -> "Leaf"
+    | Node (l, x, r) -> Printf.sprintf "Node (%s, %d, %s)" (aux l) x (aux r)
+  in
+  aux tree
+;;
+
+let () = 
+  let tree = Node (Node (Leaf, 1, Leaf), 2, Node (Leaf, 3, Leaf))
+  in
+  let result = tree_map (fun x -> x + 1) tree
+  in
+  Printf.printf "%s\n" (show_tree result)
+;;
+
+(* Suppose you have an ordered/sorted tree. Please write
+listify : ‘a tree -> ‘a list
+that creates a sorted list. *)
+
+
+let listify : 'a tree -> 'a list = 
+  let rec aux acc = function 
+  | Leaf -> acc
+  | Node (l, x, r) -> aux (x :: (aux acc r)) l
+in
+aux []
+;;
+
+let () = 
+  let tree = Node (Node (Leaf, 1, Leaf), 2, Node (Leaf, 3, Leaf))
+  in
+  let result = listify tree
+  in 
+  print_list result
+;;
+
+(* Suppose you have a sorted list. Please write the function
+treeify : ‘a list -> ‘a tree that builds a perfectly balanced sorted tree (other than perhaps a few
+missing leaves in the bottom level if there are not enough elements to make a perfectly balanced
+tree). For example: 
+let t = treeify
+[1;2;3;4;5;6;7;8;9;10;11;12;13;14;15;16;17;18;19;20;21;22;23;24;25;26;27;28;29;30;31]
+*)
+
+(* convert a sorted list to a binary tree *)
+
+type 'a treeV2 =
+  | Lf
+  | Br of 'a * 'a treeV2 * 'a treeV2
+
+let rec print_tree = function
+  | Lf -> ()
+  | Br (value, left, right) ->
+    print_string "Br (";
+    print_int value;
+    print_string ", ";
+    print_tree left;
+    print_string ", ";
+    print_tree right;
+    print_string ")"
+  ;;
+
+let list_length lst = 
+  let rec aux acc = function
+    | [] -> acc
+    | _ :: t -> aux (acc + 1) t
+in
+aux 0 lst;;
+
+let index_of lst index =
+  let rec aux acc = function
+    | [] -> 0
+    | h :: t -> if acc = index then h else aux (acc + 1) t
+in
+aux 0 lst;;
+
+(* [first:last) *)
+let partition_list l r lst =
+  let rec aux curr acc = function
+    | [] -> acc
+    | h :: t -> if curr >= l && curr < r then aux (curr + 1) (h :: acc) t else aux (curr + 1) acc t
+in
+reverse (aux 0 [] lst);;
+
+let treeify lst = 
+  let rec aux = function
+    | [] -> Lf
+    | [x] -> Br (x, Lf, Lf)
+    | lst -> 
+      let n = list_length lst in
+      let m = n / 2 in (* main node *)
+      let l = partition_list 0 m lst in
+      let r = partition_list (m+1) n lst in
+      let x = index_of lst m in
+      Br ( x, aux l, aux r)
+    in
+    aux lst;;
+
+let tree_lst = [1;2;3;4;5;6;7;8;9;10;11;12;13;14;15;16;17;18;19;20;21;22;23;24;25;26;27;28;29;30;31]
+
+let t = treeify tree_lst;;
+
+let () =
+  print_tree t;
+  print_newline ()
+  ;;
+
+let max v1 v2 = 
+  if v1 > v2 then v1 else v2
+;;
+
+let height tree_struct = 
+  let rec aux counter = function
+    | Lf -> counter
+    | Br (_, l, r) -> 
+      let left_height = aux (counter + 1) l in
+      let right_height = aux (counter + 1) r in
+      max left_height right_height
+    in
+    aux 0 tree_struct
+  ;;
```

Comment:
👍🏼, you actually don't need to define a `max` function, already exists in the Stdlib